enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,225 @@
+---
+title: remediation-history
+authors:
+  - "@slintes"
+reviewers:
+  - "@beekhof"
+approvers:
+  - "@JoelSpeed"
+  - "@michaelgugino"
+  - "@enxebre"
+creation-date: 2020-12-15
+last-updated: 2020-12-15
+status: implementable
+---
+
+# Remediation history
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Record and show limited remediation history in baremetal environments.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an *unlimited* history of remediations.
+- Collect or provide information on *why* a node has a certain node condition which triggered remediation.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was succesfully fenced (deleted / powered off)
+- when the node is healthy again.
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+MHC type enhancement:
+
+```go
+
+// MachineHealthCheckStatus defines the observed state of MachineHealthCheck
+type MachineHealthCheckStatus struct {
+
+	[...]
+
+	// History of remediations triggered by this machine health check
+	RemediationHistory []Remediation `json:"remediationHistory,omitempty"`
+}
+
+// Remediation tracks a remediation triggered by this machine health check
+type Remediation struct {
+	// the kind of the remediation target, usually node or machine
+	// +kubebuilder:validation:Type=string
+	TargetKind string `json:"targetKind"`
+
+	// the name of the machine or node which is remediated
+	// +kubebuilder:validation:Type=string
+	TargetName string `json:"targetName"`
```

Comment:
The `TargetKind` and `TargetName` fields could be combined into one of the object reference types instead of using separate fields (there are a few types, I'm not sure which is most correct). That would make it easier to support remediation outside of the default namespace, which might be useful in a multi-cluster setup.

Reply:
good idea, thanks!